Documents refresh [WIP]

SlowLogicBoy opened the pull request:
Fixes #135 
![image](https://user-images.githubusercontent.com/3121637/31437661-b0d7ce6e-ae8e-11e7-8bb0-eef53840be57.png)

Not the best solution, but it works.
Probably FileSystemWatcher should be used, and update elements accordingly, since if I refresh, all expanded directories get folded again.

TODO:

- [x] Remove refresh button, potentially leaking memory and it auto refreshes anyways
- [x] Check for dublication
- [x] Order documents?
- [x] Opened documents should track themselves too (deleted, renamed etc..)
- [x] Avalonia UI
- [x] Update subitems when renaming
- [x] Remove Empty Subscribers from DocumentWatcher
- [x] Support for static actions in DocumentWatcher?
- [x] Make sure deleted documents get collected (no mem leak)
  - [x] Unsubscribe when document root changes
- [x] DocumentWatcher to Observable pattern?
- [ ] Fix: *.autosave.csx gets added into document list.
- [ ] Fix renaming unsupported type to supported one ('example.cs' -> 'example.csx') it will not appearing in editor
- [ ] Fix: saving new document, document tree and opened document shares the same document object

Notes/Known Issues:
* Doesn't work with documents not in document root

aelij commented:
I added the word "fixes" in the description because it auto-closes the issue (see [here](https://help.github.com/articles/closing-issues-using-keywords/)).

aelij commented on src/RoslynPad.Common.UI/Services/DocumentFileWatcher.cs:
```diff
@@ -0,0 +1,114 @@
+            return new Unsubscriber<DocumentFileChanged>(_observers, observer);
+        }
+
+        ~DocumentFileWatcher() => Dispose();
```
Remove this (and `GC.SuppressFinalize` above). A finalizer is not required if the disposed object is a managed object.

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -1,46 +1,61 @@
+        public string Path
+        {
+            get => _path;
+            set
```
Make setter private

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -1,46 +1,61 @@
             {
                 Name = Name.Substring(0, Name.Length - AutoSaveSuffix.Length);
             }
-
```
Nit: restore this line (prefer an empty line after a `}`)

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -1,46 +1,61 @@
+                {
+                    foreach (var child in Children)
+                    {
+                        child.Path = child.Path.Replace (_path, value);
```
Nit: remove extra space after `Replace` (and possibly in other places) - I use the default format settings of VS

aelij commented on src/RoslynPad.Common/Utilities/IOUtilities.cs:
```diff
@@ -116,5 +116,7 @@ public static bool IsNormalIOException(Exception e)
                    e is InvalidOperationException;
         }
+
+        public static bool IsDirectory (string path) => Directory.Exists(path);
```
This method is redundant, inline it.

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -160,17 +185,80 @@ private ObservableCollection<DocumentViewModel> ReadChildren()
+                    .Where(x => !x.IsAutoSave)
+                    .OrderBy(dd => !dd.IsFolder)
+                    .ThenBy(OrderByName));
         }
```
I prefer the previous way, where directories and files were sorted separately, and it allows for directory and file objects to be built differently. Plus it saves some IO by not checking `IsDirectory` in the ctor when it's already known.

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -160,17 +185,80 @@ private ObservableCollection<DocumentViewModel> ReadChildren()
             return Regex.Replace(x.Name, "[0-9]+", m => m.Value.PadLeft(100, '0'));
         }
+
+        private static Func<IEnumerable<DocumentViewModel>, IOrderedEnumerable<DocumentViewModel>> SortPredicate =>
```
Encapsulate adding a child into a private `AddChild` method, and inline this lambda.

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -1,46 +1,61 @@
+            {
+                if (IsFolder)
+                {
+                    foreach (var child in Children)
```
This will **recursively initialize all `Children` from the ctor** - a thing we try to avoid. Normally `Children` is evaluated lazily by the UI. It would be better to leave a simple setter, and add a method that does the recursion if `_children != null` (better make a `bool IsChildrenInitialized => _children != null` and it use).

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -118,7 +133,17 @@ public bool IsExpanded
+            set
+            {
+                if (!IsFolder)
+                    value = System.IO.Path.GetFileNameWithoutExtension (value);
```
Nit: add braces to `if { }`

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -160,17 +185,80 @@ private ObservableCollection<DocumentViewModel> ReadChildren()
+
+        private void OnDocumentCreated (DocumentFileChanged value)
+        {
+            if (!IsFolder)
```
Also check `IsChildrenInitialized` - again we don't want them to be initialized by accident.

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -160,17 +185,80 @@ private ObservableCollection<DocumentViewModel> ReadChildren()
+                _documentFileWatcherDisposable.Dispose();
+                return;
+            }
+            var child = Children.SingleOrDefault(c => c.Path == value.Path);
```
Check if `IsChildrenInitialized`

aelij commented on src/RoslynPad.Common.UI/ViewModels/MainViewModelBase.cs:
```diff
@@ -141,7 +149,7 @@ public MainViewModelBase(IServiceProvider serviceProvider, ITelemetryProvider te
         {
             return IOUtilities.EnumerateFilesRecursive(root, DocumentViewModel.GetAutoSaveName("*")).Select(x =>
-                GetOpenDocumentViewModel(DocumentViewModel.FromPath(x)));
+                GetOpenDocumentViewModel(DocumentViewModel.FromPath(x, _documentFileWatcher)));
```
No need to pass the watcher here - these are just autosave docs and won't appear in the tree. Moreover - even if these files were renamed or deleted, we'd still like to save them with the original autosave name.

aelij commented on src/RoslynPad.Common.UI/ViewModels/MainViewModelBase.cs:
```diff
@@ -333,13 +341,13 @@ public void OpenDocument(DocumentViewModel document)
             }
 
-            var document = DocumentViewModel.FromPath(dialog.FileName);
+            var document = DocumentViewModel.FromPath(dialog.FileName, _documentFileWatcher);
```
Again, no need for the watcher in this case.

aelij commented on src/RoslynPad.Common.UI/ViewModels/MainViewModelBase.cs:
```diff
@@ -333,13 +341,13 @@ public void OpenDocument(DocumentViewModel document)
                 if (File.Exists(autoSavePath))
                 {
-                    document = DocumentViewModel.FromPath(autoSavePath);
+                    document = DocumentViewModel.FromPath(autoSavePath, _documentFileWatcher);
```
No need for the watcher in this case.

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -65,24 +80,24 @@ public static string GetAutoSaveName(string name)
         }
 
-        public DocumentViewModel CreateNew(string documentName)
+        public DocumentViewModel CreateNew(string documentName, DocumentFileWatcher documentFileWatcher)
```
The watcher can be accessed from the instance field, remove parameter.

aelij commented on src/RoslynPad.Common.UI/ViewModels/OpenDocumentViewModel.cs:
```diff
@@ -85,7 +91,29 @@ private set
+            }
+        }
+
+        private void DocumentOnPropertyChanged (object sender, PropertyChangedEventArgs args)
```
This is not a good editing experience IMO. If someone changes a file of a document that's currently being edited, when saving, I'd expect the original name and path to hold (that's what Notepad++ does, for example). Perhaps when opening a document, we should clone the `DocumentViewModel` and keep the clone disconnected from the watcher completely. This way the tree can update freely. 

I think you can pretty much undo all changes to this file, and add the cloning at `SetDocument`.

aelij commented on src/RoslynPad/DocumentTreeView.xaml:
```diff
@@ -27,7 +27,7 @@
                        Height="12"
                        Margin="0 0 3 0" />
-                <TextBlock Text="{Binding Name, Mode=OneTime}" />
+                <TextBlock Text="{Binding Name, Mode=OneWay}" />
```
`OneWay` is the default, can remove

aelij commented on src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs:
```diff
@@ -65,24 +80,24 @@ public static string GetAutoSaveName(string name)
         }
 
-        public static DocumentViewModel FromPath(string path)
+        public static DocumentViewModel FromPath(string path, DocumentFileWatcher documentFileWatcher)
```
I think the watcher can be removed in all usages of this method (see comments in other files).

SlowLogicBoy commented:
Thanks for review, I think I fixed most of the stuff you mentioned.

aelij commented:
Hey, is this still WIP?

SlowLogicBoy commented:
Yea, but I don't have much time to do this, since I don't have Windows Machines at home.
If you don't want this hanging you can close it :/

aelij commented:
I don't mind keeping it open. What's left to do? Maybe I'll finish it.

SlowLogicBoy commented:
Everything what's left is in Task List (because I'm terrible at remembering things), and I don't remember if I addressed your comments after last review. That would be all.

aelij commented:
Merged. I'll continue work in master. Thanks @SlowLogicBoy!
